### PR TITLE
target/riscv: resume only halted harts

### PR DESCRIPTION
**Pull request by en-sc**
With this change, failures to resume a hart due to it not being halted are more explicitly logged or reported as an error.

Change-Id: Ia55d8df85a908363d0f2140637ce1e47c1ab6251

### CODE REVIEW

File: src/target/riscv/riscv.c [hunk 1]
```diff
@@ -1880,11 +1874,21 @@ static int riscv_resume(
...
+
 	struct target_list *tlist;
 	foreach_smp_target_direction(resume_order == RO_NORMAL, tlist, targets) {
 		struct target *t = tlist->target;
+		LOG_TARGET_DEBUG(t, "target->state=%s", target_state_name(t));
 		if (t->state != TARGET_UNAVAILABLE) {
-			if (resume_prep(t, current, address, handle_breakpoints,
+			if (t->state != TARGET_HALTED)
+				LOG_TARGET_DEBUG(t, "skipping this target: target not halted");
+			else if (resume_prep(t, current, address, handle_breakpoints,
```
Comment (timsifive) [inline src/target/riscv/riscv.c]: Since this is just a debug message, you can also print it out when the hart is unavailable, simplifying the code.

**Review by timsifive — CHANGES_REQUESTED** [review]
Looks OK. I'm a little concerned about corner cases, but the spirit of this code is good. What testing did you do?

File: src/target/riscv/riscv.c [hunk 2]
```diff
@@ -1914,13 +1918,14 @@ static int riscv_resume(
...
-		target->state,
-		current ? "true" : "false",
-		address,
-		handle_breakpoints ? "true" : "false",
-		debug_execution ? "true" : "false");
+	/* FIXME: It would be better to replace this `if` with an `assert`
+	 * by moving it to the target-independent part of resume command handler.
+	 * Please, see https://review.openocd.org/c/openocd/+/7750
+	 */
+	if (target->state != TARGET_HALTED) {
```
Comment (JanMatCodasip) [inline src/target/riscv/riscv.c]: I'd recommend to leave here the "if" (instead of an assert) and drop the comment.
The "if" is preferred here for the sake of robustness - that's because we have less control over the common, target independent code, and it can change on us at some point in the future.

```suggestion
	if (target->state != TARGET_HALTED) {
```

As for https://review.openocd.org/c/openocd/+/7750 - overall it looks good but as mentioned already in the upstream review, I don't know enough about other targets to be able to confirm that change.

**Review by JanMatCodasip — DISMISSED** [review]

File: src/target/riscv/riscv.c (hunk 1 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

File: src/target/riscv/riscv.c (hunk 2 above)
Comment (en-sc) [reply src/target/riscv/riscv.c]: Addressed.

**Review by JanMatCodasip — APPROVED** [review]

**Comment by en-sc** [comment]
> Looks OK. I'm a little concerned about corner cases, but the spirit of this code is good. What testing did you do?

I've run tests from `riscv-tests/debug` testsuite on these platforms:

<details><summary>spike32</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 197s ]:::::::::::::::::::::::::::
14 tests returned not_applicable
57 tests returned pass
</details>
<details><summary>spike64</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 201s ]:::::::::::::::::::::::::::
12 tests returned not_applicable
59 tests returned pass
</details>
<details><summary>spike32-2-hwthread</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 244s ]:::::::::::::::::::::::::::
12 tests returned not_applicable
59 tests returned pass
</details>
<details><summary>spike64-2</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 265s ]:::::::::::::::::::::::::::
14 tests returned not_applicable
57 tests returned pass
</details>
<details><summary>spike32-2</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 269s ]:::::::::::::::::::::::::::
13 tests returned not_applicable
58 tests returned pass
</details>
<details><summary>spike64-2-hwthread</summary>
<details><summary> test list </summary>
CeaseMultiTest CeaseRunTest CeaseStepiTest CheckMisa CustomRegisterTest DebugBreakpoint DebugChangeString DebugCompareSections DebugExit DebugFunctionCall DebugSymbols DebugTurbostep DisconnectTest DownloadTest EbreakTest EtriggerTest FreeRtosTest Hwbp1 Hwbp2 HwbpManual IcountTest InfoTest InstantChangePc InstantHaltTest InterruptTest ItriggerTest JumpHbreak MemTest16 MemTest32 MemTest64 MemTest8 MemTestBlock0 MemTestBlock1 MemTestBlock2 MemTestBlockReadInvalid MemTestReadInvalid MemorySampleMixed MemorySampleSingle MulticoreRegTest MulticoreRtosSwitchActiveHartTest MulticoreRunAllHaltOne PrivChange PrivRw ProgramHwWatchpoint ProgramSwWatchpoint Registers RepeatReadTest Semihosting SemihostingFileio SimpleF18Test SimpleNoExistTest SimpleS0Test SimpleS1Test SimpleT0Test SimpleT1Test SimpleV13Test SmpSimultaneousRunHalt StepTest StepThread2Test Sv32Test Sv39Test Sv48Test TooManyHwbp TriggerDmode TriggerExecuteInstant TriggerLoadAddressInstant TriggerStoreAddressInstant UserInterrupt VectorTest WriteCsrs WriteGprs
</details>

:::::::::::::::::::::::::::[ ran 71 tests in 330s ]:::::::::::::::::::::::::::
12 tests returned not_applicable
59 tests returned pass
</details>

**Review by timsifive — APPROVED** [review]
